### PR TITLE
Cross-dataset: Momentum-SAM (MSAM) — flat minima without 2x compute (CODE CHANGE)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Standard SAM costs 2x compute (unacceptable at our epoch budget). Momentum-SAM (MSAM) perturbs weights in the momentum buffer direction instead — same flat-minima bias at ~0.1% overhead. At 2-11 epochs for shorter datasets, pushing toward broad basins from the first step amplifies each gradient step's value. The momentum direction is more stable than raw gradient direction, so the perturbation rho can be larger without instability. Prior work shows SAM-style methods improve generalization by finding flatter loss basins.

## Instructions

**CODE CHANGE REQUIRED:** Add `--use-msam` flag and `--msam-rho` argument (default 0.02) to `target/icml2026/train.py`. Implementation:

```python
# At each optimizer.step(), before clearing gradients:
if args.use_msam:
    with torch.no_grad():
        for group in optimizer.param_groups:
            for p in group['params']:
                if p.grad is None: continue
                state = optimizer.state[p]
                if 'exp_avg' not in state: continue  # skip if no momentum yet
                momentum = state['exp_avg']
                norm = momentum.norm()
                if norm > 0:
                    delta = args.msam_rho * momentum / norm
                    p.data.add_(delta)
        # Recompute loss and grad at perturbed point (use same batch)
        loss_perturbed = compute_loss(model, batch)
        loss_perturbed.backward()
        # Remove perturbation
        for group in optimizer.param_groups:
            for p in group['params']:
                if p.grad is None: continue
                state = optimizer.state[p]
                if 'exp_avg' not in state: continue
                momentum = state['exp_avg']
                norm = momentum.norm()
                if norm > 0:
                    delta = args.msam_rho * momentum / norm
                    p.data.sub_(delta)
    optimizer.step()
```

**Note:** This DOES require 2 forward-backward passes per step (same as regular SAM). The advantage of MSAM is that the perturbation direction (momentum) is more stable than the raw gradient direction, so rho can be larger. If this is too expensive, try with `--msam-rho 0.01` and only apply MSAM every 2nd step.

All runs use: `--no-use-ema --epochs 999 --enable-fourier`

DrivAerML extra flags: `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

TandemFoil extra flags: `--optimizer lion --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`

**Runs (use `--wandb_group msam-optimizer`):**

- **GPU 0:** TF base + `--use-msam --msam-rho 0.02`
  ```
  python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --use-msam --msam-rho 0.02 --wandb_group msam-optimizer
  ```

- **GPU 1:** AF base + `--use-msam --msam-rho 0.02`
  ```
  python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --use-msam --msam-rho 0.02 --wandb_group msam-optimizer
  ```

- **GPU 2:** DAM base + `--use-msam --msam-rho 0.02`
  ```
  python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.02 --wandb_group msam-optimizer
  ```

- **GPU 3:** DAM base + `--use-msam --msam-rho 0.05`
  ```
  python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.05 --wandb_group msam-optimizer
  ```

- **GPU 4:** DAM base + `--use-msam --msam-rho 0.01`
  ```
  python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.01 --wandb_group msam-optimizer
  ```

- **GPU 5:** DAM base + `--use-msam --msam-rho 0.02 --grad-clip 1.0`
  ```
  python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.02 --wandb_group msam-optimizer
  ```

- **GPU 6:** DAM base + `--use-msam --msam-rho 0.02 --weight-decay 1e-2`
  ```
  python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.02 --wandb_group msam-optimizer
  ```

- **GPU 7:** DAM base + `--use-msam --msam-rho 0.02 --lr 4e-4`
  ```
  python train.py --dataset drivaerml --optimizer adamw --lr 4e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.02 --wandb_group msam-optimizer
  ```

## Baseline

Current best metrics:
- **DrivAerML:** `val_primary/surface_rel_l2_pct` = **4.619%** (PR #2691, 4L/512d/8H, AdamW lr=5e-4, T_max=30). Target: 3.71%
- **AirfRANS:** `val_primary/surface_mse` = **0.001236** (PR #2828, 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5). Target: 0.0043
- **TandemFoil:** `val_primary/surface_pressure_mae` = **30.10** (PR #2810, 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT usopp:

## Results

### Summary

**MSAM is catastrophically harmful** across all three datasets, with all 8 experiments performing 3.7x to 22.7x worse than their respective baselines. No variant of rho, grad clipping, weight decay, or learning rate rescued the approach.

### TandemFoil (GPU 0) — Baseline: 30.10

| Config | Best Epoch | Best `val_primary/surface_pressure_mae` | vs Baseline |
|--------|-----------|----------------------------------------|-------------|
| MSAM rho=0.02 | 13 | **110.25** | 3.7x worse |

### AirfRANS (GPU 1) — Baseline: 0.001236 (target: 0.0043)

| Config | Best Epoch | Best `val_primary/surface_mse` | vs Baseline |
|--------|-----------|-------------------------------|-------------|
| MSAM rho=0.02 | 60 | **0.028005** | 22.7x worse |

### DrivAerML (GPUs 2-7) — Baseline: 4.619% (target: 3.71%)

| GPU | Config | Best Epoch | Best `val_primary/surface_rel_l2_pct` | vs Baseline |
|-----|--------|-----------|--------------------------------------|-------------|
| 2 | rho=0.02 (base) | 33 | **44.60%** | 9.7x worse |
| 3 | rho=0.05 | 47 | **59.85%** | 13.0x worse |
| 4 | rho=0.01 | 20 | **34.30%** | 7.4x worse |
| 5 | rho=0.02 + grad_clip=1.0 | 47 | **29.84%** | 6.5x worse |
| 6 | rho=0.02 + weight_decay=1e-2 | 36 | **48.40%** | 10.5x worse |
| 7 | rho=0.02 + lr=4e-4 | 40 | **40.17%** | 8.7x worse |

### Commands

```bash
# GPU 0 (TF)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --use-msam --msam-rho 0.02 --agent usopp --wandb-name "usopp/tf-msam-rho0.02" --wandb-group "msam-optimizer"

# GPU 1 (AF)
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --use-msam --msam-rho 0.02 --agent usopp --wandb-name "usopp/af-msam-rho0.02" --wandb-group "msam-optimizer"

# GPUs 2-7 (DAM variants)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --use-msam --msam-rho 0.02 --agent usopp --wandb-name "usopp/dam-msam-rho0.02" --wandb-group "msam-optimizer"
# (GPU 3: rho=0.05, GPU 4: rho=0.01, GPU 5: +grad-clip 1.0, GPU 6: +weight-decay 1e-2, GPU 7: lr=4e-4)
```

### Peak Memory

- TandemFoil (GPU 0): ~36 GB
- AirfRANS (GPU 1): ~31 GB
- DrivAerML (GPUs 2-7): ~6 GB each

### W&B Run IDs

- GPU 0 (TF): `gtmy1tv6`
- GPU 1 (AF): `yix79ft2`
- GPU 2 (DAM rho=0.02): `1fblqqcn`
- GPU 3 (DAM rho=0.05): `nzdchctr`
- GPU 4 (DAM rho=0.01): `j0zlgckv`
- GPU 5 (DAM rho=0.02+gc): `f4fbccfp`
- GPU 6 (DAM rho=0.02+wd): `aart20yt`
- GPU 7 (DAM rho=0.02+lr): `nkbighnc`

W&B group: `msam-optimizer`

### What happened

MSAM failed catastrophically across all datasets. The approach is fundamentally incompatible with our short-epoch training regime:

1. **Double compute per step**: Each MSAM step requires 2 forward-backward passes. With fixed epoch budgets, this halves the effective number of useful gradient updates. For our already-tight training budgets (2-11 epochs for shorter datasets, ~60 for DrivAerML), losing half the training signal is devastating.

2. **DrivAerML instability without grad clipping**: GPU 2 (no grad clipping) showed grad norm spikes up to 277, causing wild oscillations in val_primary (44-67% range). The momentum perturbation amplifies instability in the gradient landscape.

3. **Grad clipping helps but doesn't save it**: GPU 5 (best DrivAerML variant, with grad_clip=1.0) achieved 29.84% — still 6.5x worse than the 4.619% baseline. Grad clipping tamed the instability but can't compensate for halved training efficiency.

4. **Negative msam_loss_increase on TandemFoil**: TandemFoil consistently showed *negative* msam_loss_increase (~-0.01 to -0.02), meaning the momentum perturbation actually found *lower* loss points, not higher ones. This defeats SAM's purpose of finding flat minima — the perturbation is moving downhill rather than finding adversarial directions. This likely happens because the Lion optimizer's momentum buffer points in a direction that doesn't align with the local ascent direction.

5. **rho sensitivity on DrivAerML**: Higher rho (0.05) → worse. Lower rho (0.01) → better than 0.02 but still terrible. The perturbation magnitude doesn't rescue the fundamental problem.

### Suggested follow-ups

- **SAM-style optimizers are likely a dead end for our regime.** The 2x compute cost per step is unacceptable at our tight epoch budgets. Any variant that requires an extra forward-backward pass will face the same fundamental problem.
- If flat minima are desired, **stochastic weight averaging (SWA)** would be cheaper — it just averages model weights over the last portion of training, adding zero per-step overhead.
- **Gradient noise injection** (adding small Gaussian noise to gradients) could provide a lightweight flat-minima bias without the extra forward-backward pass cost.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2904 because ADVISOR: Closing — MSAM is a clear dead end across all three datasets:

| Dataset | Best MSAM | Baseline | Ratio |
|---|---|---|---|
| TandemFoil | 110.25 | 30.10 | 3.7x worse |
| AirfRANS | 0.0256 | 0.001236 | 20.7x worse |
| DrivAerML | 29.84% (rho=0.02+gc) | 4.619% | 6.5x worse |

**Three root causes you correctly diagnosed:**
1. The efficiency claim was false — implementation is 2x compute, same as standard SAM
2. Lion's momentum buffer doesn't align with loss ascent direction (negative msam_loss_increase on TF — sharp observation)
3. Perturbation amplifies gradient instability without gc

Your analysis was excellent — the negative msam_loss_increase finding on TandemFoil is a genuinely useful insight about Lion's momentum semantics. The `_forward_and_loss()` refactor is clean code.

Your suggested follow-ups (SWA, gradient noise) are good — sukuna #2903 is already testing SWA. I'll assign you gradient noise injection as your next experiment..
